✨ feat: 내가 지원한 내역 조회 API 구현

Jiwoong0322 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #198 
- 작업 요약: 내가 지원한 내역 조회 API를 개발하고, 공고 삭제 정책을 요구사항에 맞게 수정했습니다.

## 📄 상세 내용
  - [ ] View-Service-Manager 아키텍처 패턴 적용
  - [ ] 커서 기반 페이지네이션 및 N+1 문제 최적화 적용
  - [ ] Application 모델의 on_delete 정책을 요구사항에 맞게 `SET_NULL에서 CASCADE`로 수정
  - [ ] 모델 변경에 따른 마이그레이션 파일 생성 및 관련 테스트 코드 수정
  - [ ] S3 업로드를 모방(moto)하고 엣지 케이스(공고 삭제)를 포함한 단위 테스트 코드 작성


## 📸 스크린샷 (선택)

## 📝 기타 참고 사항
- 공고 삭제 시(on_delete), 관련 지원 내역(Application)을 남기는 SET_NULL 방식이 요구사항과 일치하지 않는 점을 발견하여, 함께 삭제되는 CASCADE 방식으로 수정하고 관련 테스트를 보강했습니다.

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Jiwoong0322 committed:
✨ feat: 내가 지원한 내역 조회 API 구현

내가 지원한 내역 조회 API 구현 (GET /api/v1/applications/me)
- 커서 기반 페이지네이션(Cursor Pagination) 기능 적용
- N+1 문제 방지를 위해 select_related, prefetch_related 적용
- 연관된 공고가 삭제된 경우 on_delete=SET_NULL)에도 500 에러 없이 안전하게 조회되도록 시리얼라이저 수정
- S3 업로드를 모방(moto)하고 엣지 케이스를 포함한 단위 테스트 코드 작성
- Application 모델의 on_delete 정책을 `SET_NULL에서 CASCADE`로 변경하여, 공고 삭제 시 관련 지원 내역이 함께 삭제되도록 요구사항에 맞게 수정
- 변경된 정책을 DB에 반영하기 위한 마이그레이션 파일을 생성하고, 새 정책을 검증하도록 테스트 코드 수정

관련 이슈: #198

# Conflicts:
#	apps/applications/urls/applications_urls.py

Umdoong commented on apps/applications/models/applications.py:
```diff
@@ -17,8 +17,7 @@ class ApplicationStatus(models.TextChoices):
         Recruitment,
-        on_delete=models.SET_NULL,
-        null=True,  # , or IntegrityError
+        on_delete=models.CASCADE,
```
ERD 상으로 set null로 되어있는데 변경하신 이유가 있을까요?

Jiwoong0322 replied:
요구사항 정의서에 공고가 삭제되면 데이터베이스의 해당 공고의 지원 내역들도 모두 삭제된다고 되어있어 Set NULL보다는 CASCADE를 쓰는게 맞다고 생각했는데 ERD를 재검토할 생각을 미처 하지 못했습니다. ERD에 set null로 명시되어있으니 set null을 유지하며 삭제된 공고의 지원 내역들을 데이터베이스에서 삭제하는 방법으로 수정하겠습니다.

Meoyoug replied:
이거 제가 ERD업데이트를 안한탓입니다..죄송합니다

Umdoong commented on apps/applications/serializers/applications_list_serializers.py:
```diff
@@ -33,3 +41,57 @@ class Meta:
         ]
+
+
+class MyApplicationListSerializer(serializers.ModelSerializer[Application]):
```
내가 지원한 공고 내역을 조회할 때 사용하는 거니까 model을 recruitment로 하는 게 맞지 않을까용...?

Meoyoug replied:
내가 지원한 내역을 보는거라 application이 주가 되어야하는ㄱ ㅔ맞아요

Jiwoong0322 committed:
♻️ refactor: 피드백 적용 및 마이그레이션 되돌리기 실행

  - '내가 지원한 내역 조회' API(REQ-APLY-006)의 기반 모델을 Recruitment로 변경하고, Subquery를 사용해 성능을 개선
  - ERD에 따라 Application 모델의 on_delete를 SET_NULL로 변경하고 마이그레이션을 0004에서 0003으로 되돌리기 실행
  - on_delete 정책과 별개로 공고 삭제(REQ-RECM-009) 시 지원 내역이 함께 삭제되도록 View의 삭제 로직을 수정함
  - 변경된 삭제 로직을 검증하는 API 테스트를 추가하고, 불필요한 기존 테스트는 삭제함

관련 이슈: #198

Jiwoong0322 committed:
♻️ refactor: CASCADE 재적용 및 마이그레이션, 시리얼라이저와 뷰 리팩터링

- Application 모델의 on_delete 정책을 SET_NULL에서 CASCADE로 되돌리고, 이에 대한 마이그레이션 파일을 생성
- RecruitmentDetailView에서 불필요해진 커스텀 삭제 로직을 제거
- '내가 지원한 내역 조회' API의 MyApplicationListSerializer와 MyApplicationsListView를 Application 모델 기반으로 재구성
- SerializerMethodField 사용을 최소화하고 source 매핑을 활용하여 시리얼라이저를 리팩터링

관련 이슈: #198
